### ISSUE TITLE
radix: add a "never" option

### ISSUE THREAD

**Issue by mzgol** [eslint/eslint#4048]
If one targets modern browsers only or has a transpilation step that transforms radix-less `parseInt` into one compatible with older browsers it might be undesirable to enter the radix if it's statically determinable to be equal to 10. It would be useful if the `radix` rule had accepted an option: `"always"` or `"never"` (`"always"` by default) and in the `"never"` mode would forbid entering such a radix.

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by nzakas** [comment]
In that case, wouldn't you just turn the rule off?

**Comment by silverwind** [comment]
+1 for a rule like this.

@nzakas `parseInt("071", 10)` should be an error, which could only be achieved when the rule is enabled.

**Comment by mzgol** [comment]
Yup, I'd like `"radix": [2, "never"]` to reject `parseInt(sth, 10)`. Ideally it'd also reject things like:
```js
function parseIntWrapped(x) {
    const radix = 10;
    return parseInt(x, radix);
}
```

but such cases are not common in my experience so even just rejecting the constant `10` as a second param would be enough for me.

**Comment by nzakas** [comment]
OK. Any interest in implementing yourself? It's a fairly small change.

**Comment by mzgol** [comment]
Yup, I can do it tomorrow or the day after that.

**Comment by mzgol** [comment]
PR: #4084

**Comment by ljharb** [comment]
This doesn't appear to be added to the documentation for `radix`. When is "as-needed" considered "needed"?

**Comment by ilyavolodin** [comment]
@ljharb Thanks for noticing. I opened #4364 to add documentation for `as-needed` option. In short `as-needed` option will flag all usage of radix when it's equal to 10, as it is the default value. So an expression like this will be flagged:
```js
parseInt(someVar, 10);
```

**Comment by ljharb** [comment]
So then `parseInt('0x5', 16)` and `parseInt('07', 8)` and `parseInt('0b1', 2)` and `parseInt('0o7', 8)` would also be flagged when "as-needed" is enabled?

**Comment by ljharb** [comment]
(In ES5+, `10` is only the default value if the string is not prefixed by `0x` or `0X`)

**Comment by mzgol** [comment]
@ljharb Let me start from the latter examples:

1. `parseInt('07', 8)` - do you mean that `8` is redundant here because we only have one digit? That seems like a stretch, `parseInt('077') !== parseInt('077', 8)` so I wouldn't mark this specific one as `not-needed`.
2. `parseInt('0b1', 2)` - this is not valid, `0b1` will parse as `0` regardless of the radix passed. Same for `parseInt('0o7', 8)`; the only valid prefix form is `0xSOMEDIGITS`.

As for `parseInt('0x5', 16)` - a valid point that the `radix` is not needed there but we're not talking constants here but a form:
```js
parseInt(someVar, 16)
```
where you can't always know what `someVar` holds. Whereas in the case of:
```js
parseInt(someVar, 10)
```
the radix can be skipped as the only difference in the semantics of this line is when `someVar` starts with `0x` or `0X` but then `parseInt(someVar, 10)` would return 0 which most likely isn't the intended scenario. If it was done on purpose, configuring `"radix": [2, "as-needed"]` would mean in such a case you'd need to rewrite this line to handle those prefixes explicitly.

Do you have another proposal for how this rule should behave?

**Comment by mzgol** [comment]
Docs PR: #4367

**Comment by ljharb** [comment]
> would return 0 which most likely isn't the intended scenario

I think I'm just questioning the value of this rule by exploring specific examples - the "most likely" is the same uncertainty that makes "always include the radix" a best practice, and I don't think ES6 changes that. However, it's not the default, so as long as it's documented, I should probably bow out here :-)
